subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/ComponentMetadataRulesCachingIntegrationTest.groovy

```diff
@@ -194,4 +196,168 @@ dependencies {
         outputContains('Rule B executed - saw changing true')
     }
 
+    def 'can cache rules having a custom type attribute as parameter'() {
+        repository {
+            'org.test:projectA:1.0'()
+        }
+        buildFile << """
+
+import javax.inject.Inject
+
+@CacheableRule
+class AttributeCachedRule implements ComponentMetadataRule {
+
+    Attribute targetAttribute
+
+    @Inject
+    AttributeCachedRule(Attribute attribute) {
+        this.targetAttribute = attribute
+    }
+    
+    void execute(ComponentMetadataContext context) {
+        println 'Attribute rule executed'
+    }
+}
+
+class Thing implements Named, Serializable { 
```

Comment:
:x: This should be changed to use `interface Thing extends Named`, and use the object factory instead. Can be done after merging in a polishing step.